Support for multiple fields on IBM Power Virtual Servers provisioning form.

iv1111 opened the pull request:
- added support for the user post-provisioning script upload
- added support for multiple volume attachment to provisioning form
- added support for "IP-Address v4" field to provisioning form
- added CPU amount validation based on the CPU type on provisioning form
- changed Memory field from drop-down to text-input on provisioning form
- added "Keypair" field to the provisioning form
- added support for "VM Pinning", "Migratable", "System Type" to the provisioning form
- added support for 'Entitled Processors' to the provisioning form
- moved 'Storage Type' and 'System Type' values to the provision_workflow code
- using getter for the access to EMS data in DB in provision_workflow
- fix in a spec containing a wrong AIX OS name
- fix for the case of empty set of values in the 'volumes' field

agrare commented:
Hey @iv1111 it looks like you have some commits from master in here, make sure you `git pull --rebase` or (preferably) have `git config pull.rebase true`

agrare commented:
Also @iv1111 what is `Issue #82` :wink:

iv1111 commented:
> Hey @iv1111 it looks like you have some commits from master in here, make sure you `git pull --rebase` or (preferably) have `git config pull.rebase true`

Don't understand why this happens. What I usually do is either:

-  `git fetch manageiq`
- `git rebase manageiq/master`

or through by having rebase as default:

- `git config pull.rebase true`

agrare commented:
@iv1111 if you have `pull.rebase true` then all you need to do is `git pull manageiq/master` (assuming `manageiq` is the name of the upstream remote).  `git pull` will automatically fetch for you so you don't need to do that in a separate step

iv1111 commented:
Ok. I'll make sure this doesn't happen again.

jaywcarman commented on content/miq_dialogs/miq_provision_ibm_dialogs_template.yaml:
```diff
@@ -13,9 +13,46 @@
+      :description: Customization
+      :fields:
+        :ip_addr:
+          :description: IP v4 Address
```
If possible, we should try and match these descriptions to what is in the web UI:
"Specify an IPv4 address"

![image](https://user-images.githubusercontent.com/1637291/92963185-ad887000-f437-11ea-89ab-9f629210da0d.png)

jaywcarman commented on content/miq_dialogs/miq_provision_ibm_dialogs_template.yaml:
```diff
@@ -13,9 +13,46 @@
+            1: "none"
+            2: "soft"
+            3: "hard"
+          :description: Pin Policy
```
"VM pinning"
![image](https://user-images.githubusercontent.com/1637291/92963282-d6106a00-f437-11ea-80dc-9b0eec58dda9.png)

jaywcarman commented on content/miq_dialogs/miq_provision_ibm_dialogs_template.yaml:
```diff
@@ -61,39 +98,64 @@
-            8: 8
-          :description: Cores (CPUs)
+        :entitled_processors:
+          :description: Entitled CPUs
```
"Cores (Entitled CPUs)"
![image](https://user-images.githubusercontent.com/1637291/92963477-225baa00-f438-11ea-99ae-d0f954e217bd.png)

jaywcarman commented on content/miq_dialogs/miq_provision_ibm_dialogs_template.yaml:
```diff
@@ -61,39 +98,64 @@
+        :guest_access_key_pair:
+          :values_from:
+            :method: :allowed_guest_access_key_pairs
+          :description: Key Pair
```
The web UI uses "SSH Key", but in this case maybe it's best to keep "Key Pair" to stay consistent with ManageIQ's label?
![image](https://user-images.githubusercontent.com/1637291/92963944-e543e780-f438-11ea-8f87-687bdaa251d6.png)

iv1111 replied:
we could keep "Key Pair" for now, it is easy to change anyway

jaywcarman reviewed (commented):
It's great to see this coming together @iv1111 !

In preparation of the incoming VPC code, should we rename `miq_provision_ibm_dialogs_template.yaml` to something like `miq_provision_ibm_cloud_power_virtual_servers_dialogs_template.yaml` ? (or something shorter, but still indicating that it is specific to PowerVS)

iv1111 commented on app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager/provision_workflow.rb:
```diff
@@ -16,16 +18,70 @@ def get_timezones(_options = {})
+  def validate_entitled_processors(field, values, dlg, fld, value)
+    dedicated = values[:instance_type][1] == 'dedicated'
+
+    fval = value.match(/^\s*[\d]+(\.[\d]+)?\s*$/) ? value.strip.to_f : 0
```
Ok, I'll replace the first `[\d]+` with` [\d]*` then.

iv1111 commented on app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager.rb:
```diff
@@ -17,15 +15,21 @@ class ManageIQ::Providers::IbmCloud::PowerVirtualServers::CloudManager < ManageI
+
   before_create :ensure_managers
-  before_create :ensure_managers_zone
+  before_update :ensure_managers_zone
```
ok, let me try and it should hopefully disappear

iv1111 commented on app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager/provision_workflow.rb:
```diff
@@ -16,16 +18,70 @@ def get_timezones(_options = {})
+    dedicated = values[:instance_type][1] == 'dedicated'
+
+    fval = value.match(/^\s*[\d]+(\.[\d]+)?\s*$/) ? value.strip.to_f : 0
+    return "Entitled Processors field does not contain a well-formed positive number" unless fval > 0
```
Yeah, I Joe told me to use "Entitled Processors" as a description, I will change the YAML.

iv1111 commented:
* fixed most of the coding style suggestions, where possible
* made changes according to Jay's suggestions
* adjusted the PR message and subject

If it's Ok then let's merge this by squashing.

iv1111 commented:
**Way to invoke IBM provisioning form:**
![Provision](https://user-images.githubusercontent.com/69972357/93768107-3615c600-fc19-11ea-997b-eb9de98f7f4b.png)


-----


**No visual Changes:**
![General_old](https://user-images.githubusercontent.com/69972357/93768106-3615c600-fc19-11ea-82cb-a5ea92f71428.png)
![General_new](https://user-images.githubusercontent.com/69972357/93768101-357d2f80-fc19-11ea-9136-521731557f36.png)



-----



**Changes present:**
![Profile_old](https://user-images.githubusercontent.com/69972357/93768105-357d2f80-fc19-11ea-97c2-4369c2ed8471.png)
![Profile_new](https://user-images.githubusercontent.com/69972357/93768100-34e49900-fc19-11ea-9d49-76cb33a650da.png)

-----



**No visual Changes:**
![Network_old](https://user-images.githubusercontent.com/69972357/93768103-357d2f80-fc19-11ea-956c-17c849a529ee.png)
![Network_new](https://user-images.githubusercontent.com/69972357/93768099-34e49900-fc19-11ea-8879-75e57f637cd4.png)


-----



**New Tab:**
![Customization_new](https://user-images.githubusercontent.com/69972357/93768095-33b36c00-fc19-11ea-9c67-841ec91d07a7.png)

jaywcarman commented:
@iv1111 and I don't see any obvious benefit to refactoring w.r.t. the codeclimate fail https://codeclimate.com/github/ManageIQ/manageiq-providers-ibm_cloud/pull/19

@agrare Are you OK to merge as-is? (@iv1111 was going to squash first)

agrare commented:
I agree that's not an overly complicated method, LGTM
